packages/wasm-crypto/src/rs/vrf.rs

```diff
@@ -0,0 +1,211 @@
+//
+// Originally developed (as a fork) in https://github.com/polkadot-js/schnorrkel-js/
+// which was adpated from the initial https://github.com/paritytech/schnorrkel-js/
+// forked at commit eff430ddc3090f56317c80654208b8298ef7ab3f
+
+use merlin::Transcript;
+use schnorrkel::{
+    signing_context,
+    vrf::{VRFOutput, VRFProof},
+    PublicKey, SecretKey,
+};
+use wasm_bindgen::prelude::*;
+
+/// Size of VRF output, bytes
+pub const VRF_OUTPUT_SIZE: usize = 32;
+
+/// Size of VRF proof, bytes
+pub const VRF_PROOF_SIZE: usize = 64;
+
+/// Size of VRF raw output, bytes
+pub const VRF_RAW_OUTPUT_SIZE: usize = 16;
+
+/// Size of VRF limit, bytes
+pub const VRF_THRESHOLD_SIZE: usize = 16;
+
+/// Run a Random Verifiable Function (VRF) on one single input (message) transcript,
+/// producing the output signature and corresponding short proof.
+///
+/// * secret: UIntArray with 64 element
+/// * context: Arbitrary length UIntArray
+/// * message: Arbitrary length UIntArray
+///
+/// * returned vector is the 32-byte output (signature) and 64-byte proof.
+#[wasm_bindgen]
+pub fn ext_vrf_sign(secret: &[u8], context: &[u8], message: &[u8]) -> Vec<u8> {
+    let keypair = SecretKey::from_ed25519_bytes(secret).unwrap().to_keypair();
+
+    let (io, proof, _) = keypair.vrf_sign(signing_context(context).bytes(message));
+
+    let mut result: [u8; VRF_OUTPUT_SIZE + VRF_PROOF_SIZE] =
+        [0u8; VRF_OUTPUT_SIZE + VRF_PROOF_SIZE];
+    result[..VRF_OUTPUT_SIZE].copy_from_slice(io.as_output_bytes());
+    result[VRF_OUTPUT_SIZE..].copy_from_slice(&proof.to_bytes());
+    result.to_vec()
+}
+
+/// Verify VRF proof for one single input transcript and corresponding output.
+///
+/// * pubkey: UIntArray with 32 element
+/// * context: Arbitrary length UIntArray
+/// * message: Arbitrary length UIntArray
+/// * out_and_proof: 96-byte output & proof array from the ext_vrf_sign function.
+#[wasm_bindgen]
+pub fn ext_vrf_verify(pubkey: &[u8], context: &[u8], message: &[u8], out_and_proof: &[u8]) -> bool {
+    match (
+        PublicKey::from_bytes(pubkey),
+        VRFOutput::from_bytes(&out_and_proof[..VRF_OUTPUT_SIZE]),
+        VRFProof::from_bytes(&out_and_proof[VRF_OUTPUT_SIZE..(VRF_OUTPUT_SIZE+VRF_PROOF_SIZE)]),
+    ) {
+        (Ok(public), Ok(out), Ok(proof)) => public
+            .vrf_verify(signing_context(context).bytes(message), &out, &proof)
+            .is_ok(),
+        _ => false,
+    }
+}
+
+/// Run a Random Verifiable Function (VRF) on one single input
+/// (message) transcript, and an extra message transcript,
+/// producing the output signature and corresponding short proof.
+///
+/// * secret: UIntArray with 64 element
+/// * context: Arbitrary length UIntArray
+/// * message: Arbitrary length UIntArray
+/// * extra: Arbitrary length UIntArray
+///
+/// * returned vector is the 32-byte output (signature) and 64-byte proof.
+#[wasm_bindgen]
+pub fn ext_vrf_sign_extra(secret: &[u8], context: &[u8], message: &[u8], extra: &[u8]) -> Vec<u8> {
```

Comment:
imho too many options with sign/sign_extra (generally prefer a single interface so there is no confusion from a user perspective). So for instance when using would always use the extra, but in the case of it not being available would pass an empty array.

But those changes that will just require ripping out, will adjust as it hits common and it is wrapped.

Reply:
That's a good point, but I would suggest getting the input of @burdges on that point, as there is a clear separation (between the two functions) in [schnorrkel](https://github.com/w3f/schnorrkel/blob/8fa2ad3e9fbf0b652c724df6a87a4b3c5500f759/src/vrf.rs#L657).
Also, I am not sure about using the same hardcoded "VRF" context (as in schnorrkel) [here](https://github.com/polkadot-js/wasm/blob/a7f83424586c0148831977c9aa5e361e3f4685f1/packages/wasm-crypto/src/rs/vrf.rs#L85) and [here](https://github.com/polkadot-js/wasm/blob/a7f83424586c0148831977c9aa5e361e3f4685f1/packages/wasm-crypto/src/rs/vrf.rs#L120).

Reply:
The transcript is the same with `VRF` as in the sr25519 base, so it aligns - it means that implementations will (hopefully) be cross-compatible.